runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -209,7 +209,12 @@ public void processElement(WindowedValue<InputT> compressedElem) {
         break;
       case PROCESSING_TIME:
       case SYNCHRONIZED_PROCESSING_TIME:
-        effectiveTimestamp = stepContext.timerInternals().currentInputWatermarkTime();
+        Instant outputWatermark = stepContext.timerInternals().currentOutputWatermarkTime();
+        Instant inputWatermark = stepContext.timerInternals().currentInputWatermarkTime();
+        effectiveTimestamp =
+            outputTimestamp != null
+                ? outputTimestamp
+                : outputWatermark != null ? outputWatermark : inputWatermark;
```

Comment:
Should the default be `BoundedWindow.TIMESTAMP_MIN_VALUE`? I also think this should be `min(window.maxTimestamp, outputWatermark)`.

Reply:
I really don't even think there should be a default, afaik `outputTimestamp` is always non-null.  It's set here in [TimerInternalsTimer](https://github.com/apache/beam/blob/master/runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java#L1248), although its possible a runner implementation doesn't correctly round-trip this information at firing time.  I'm still curious why it's like this, my guess is that it was from before timers had output timestamps, and was just never updated when they were introduced.

imo there should be little validation here, by the time the timer has fired the output timestamp should have been validated, although a check such as `assert(!outputTimestamp.isAfter(window.maxTimestamp())` would have been useful to catch this much earlier.

Reply:
agreed, this chooses outputTimestamp, output watermark, input watermark, in that order.  In practice I've never seen output timestamp not set, should I change this to use only output timestamp and assert it's non-null?

Reply:
Yes, I was referring to the case when `outputTimestamp` is null (for whatever reason). Is processing time `outputTimestamp` guaranteed to be inside the window?

Reply:
It should be, yes.  It's either the timestamp of the element that set the timer (which should always be inside the window, unless bugs like this get in :smile: ) or a user-specified timestamp which is validated to also be in the window (plus allowed lateness).  See `TimerInternalsTimer.setAndVerifyOutputTimestamp` for the validation details.

Reply:
I'm confused. There is a comment in the mentioned `setAndVerifyOutputTimestamp`, that states:
```java
     * <ul>
     *   Ensures that:
     *   <li>Users can't set {@code outputTimestamp} for processing time timers.
     *   <li>Event time timers' {@code outputTimestamp} is set before window expiration.
     * </ul>
```

The method does not check the first part apparently. If this is correct we might want to fix the comment.

Reply:
This comment just seems out of date, you certainly can set the output timestamp for processing time timers.

Reply:
I updated the comments and cleaned up the logic in `setAndVerifyOutputTimestamp`, I think it should be much clearer how its supposed to work now.

Reply:
I'm still not sure why outputTimestamp would ever be null! Does flink set it to null?

Reply:
I've never seen it actually be null, I'm happy to simplify all of this if we're assuming it can't be.

Reply:
done

Reply:
@reuvenlax can you check this out again?  It's much more simple now w/o the explicit need to handle a null case.